### PR TITLE
Avoid SIGSEGV in tromp solver

### LINKED ISSUE #555: komodod crash in mining code
With mining on, komodod sometimes crashes (observed on a dedicated node, with no ASIC).
Here is the backtrace:
```
Core was generated by `./komodod -gen -genproclimit=1 -pubkey=...'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  BitcoinMiner (pwallet=<optimized out>) at miner.cpp:1624
1624                                    index_vector[i] = eq.sols[s][i];
[Current thread is 1 (Thread 0x7f442bfff700 (LWP 26388))]
```
It looks like that the index 's' ranging from 0 from eq.nsols may get over eq.sols upper bound equal to MAXSOLS.
To fix this, the suggestion is to have the range for 's' from 0 to min(MAXSOLS, eq.nsols)

### PR DESCRIPTION
**Pull request by jmjatlanta**
Fixes #555 

The tromp equihash solver can sometimes get in the state whereby the nsols counter is greater than MAXSOLS. An attempt is then made to read an element of the array beyond the array size (index out of bounds). This fix limits the reads to a maximum of MAXSOLS.

TODO: 
- [x] Build a unit test that verifies the problem and the fix.

### CODE REVIEW

**Comment by dimxy** [comment]
looks good

File: src/miner.cpp [hunk 1]
```diff
@@ -1616,21 +1637,8 @@ void static BitcoinMiner()
...
-                            }
-                            std::vector<unsigned char> sol_char = GetMinimalFromIndices(index_vector, DIGITBITS);
-
-                            if (validBlock(sol_char)) {
-                                // If we find a POW solution, do not try other solutions
-                                // because they become invalid as we created a new block in blockchain.
-                                break;
-                            }
-                        }
+                        check_tromp_solution(eq, validBlock);
```
Comment (DeckerSU) [inline src/miner.cpp:1640]: May be here we need:
```
if (check_tromp_solution(eq, validBlock))
    break;
```
To preserve the original logic? If we find a POW solution, do not try other solutions.

Comment (dimxy) [reply src/miner.cpp:1640]: As I can tell, check_tromp_solution does not change the original logic: 
the original logic breaks the loop if a solution is found
and check_tromp_solution in this case just returns also breaking solution searching.
It looks like the boolean return is needed for tests only

Comment (DeckerSU) [reply src/miner.cpp:1640]: Yes, seems you are right. The logic is fine.

p.s.  It was "small display issue" (I mixed up the scopes somehow and what it breaks).

**Review by ca333 — APPROVED** [review]

**Comment by DeckerSU** [comment]
LGTM.

- also propagated to Komodo-Qt - [Avoid SIGSEGV in tromp solver](https://github.com/DeckerSU/KomodoOcean/commit/1bccb82383fcafd71903292f617fac7e4e46c882)
- and created corresponding PR to zcash upstream - [miner: fix MAXSOLS #6114](https://github.com/zcash/zcash/pull/6114)

**Review by DeckerSU — APPROVED** [review]

**Comment by tonymorony** [comment]
these changes implemented in combined PR: https://github.com/KomodoPlatform/komodo/pull/559

**Pull request closed without merging**